File: src/main/java/mods/cybercat/gigeresque/common/block/AcidBlock.java
package mods.cybercat.gigeresque.common.block;

import mods.cybercat.gigeresque.client.particle.Particles;
import mods.cybercat.gigeresque.common.status.effect.GigStatusEffects;
import mods.cybercat.gigeresque.common.tags.GigTags;
import mods.cybercat.gigeresque.common.util.MathUtil;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.util.RandomSource;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.*;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;

public class AcidBlock extends FallingBlock implements SimpleWaterloggedBlock {
    private static final int MAX_THICKNESS = 4;
    public static final IntegerProperty THICKNESS = IntegerProperty.create("thickness", 0, MAX_THICKNESS);
    private static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
    private int age = 0;

    AcidBlock(Properties settings) {
        super(settings);
        registerDefaultState(
                (getStateDefinition().any().setValue(WATERLOGGED, false)).setValue(THICKNESS, MAX_THICKNESS));
    }

    private void scheduleTickIfNotScheduled(Level world, BlockPos pos) {
        if (!world.isClientSide && !world.getBlockTicks().hasScheduledTick(pos, this)) {
            age++;
            world.scheduleTick(pos, this, 40);
        }
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(THICKNESS, WATERLOGGED);
    }

    @Deprecated(since = "1.20")
    @Override
    public void onPlace(@NotNull BlockState state, Level world, @NotNull BlockPos pos, @NotNull BlockState oldState, boolean notify) {
        if (age > MathUtil.clamp(world.random.nextInt(2) + 1, 1, 52)) {
            world.setBlock(pos, Blocks.AIR.defaultBlockState(), 2);
            age = -1;
        }
        scheduleTickIfNotScheduled(world, pos);
    }

    @Override
    public boolean isRandomlyTicking(@NotNull BlockState state) {
        return true;
    }

    @Override
    public float getExplosionResistance() {
        return 100.0f;
    }

    @Deprecated(since = "1.20")
    @Override
    public @NotNull RenderShape getRenderShape(@NotNull BlockState state) {
        return RenderShape.INVISIBLE;
    }

    @Deprecated(since = "1.20")
    @Override
    public boolean isPathfindable(@NotNull BlockState state, @NotNull BlockGetter world, @NotNull BlockPos pos, @NotNull PathComputationType type) {
        return true;
    }

    @Deprecated(since = "1.20")
    @Override
    public @NotNull FluidState getFluidState(BlockState state) {
        return Boolean.TRUE.equals(state.getValue(StairBlock.WATERLOGGED)) ? Fluids.WATER.getSource(
                false) : super.getFluidState(state);
    }

    @Deprecated(since = "1.20")
    @Override
    public @NotNull BlockState updateShape(@NotNull BlockState state, @NotNull Direction direction, @NotNull BlockState neighborState, LevelAccessor world, @NotNull BlockPos pos, @NotNull BlockPos neighborPos) {
        if (!world.isClientSide() && Boolean.TRUE.equals(state.getValue(WATERLOGGED)))
            world.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(world));
        return super.updateShape(state, direction, neighborState, world, pos, neighborPos);
    }

    private int getThickness(BlockState state) {
        return state.getValue(THICKNESS);
    }

    @Override
    public void playerWillDestroy(@NotNull Level world, @NotNull BlockPos pos, @NotNull BlockState state, @NotNull Player player) {
        // Stop Vanilla Stuff
    }

    private void setThickness(ServerLevel world, BlockPos pos, BlockState state, int consume) {
        var newState = state.setValue(THICKNESS, Math.max(getThickness(state) - consume, 0));

        if (world.getBlockState(pos).getBlock() == Blocks.WATER) newState = newState.setValue(WATERLOGGED, true);

        world.setBlockAndUpdate(pos, newState);
    }

    @Override
    public void playerDestroy(@NotNull Level world, @NotNull Player player, @NotNull BlockPos pos, @NotNull BlockState state, BlockEntity blockEntity, @NotNull ItemStack stack) {
        // Stop Vanilla Stuff
    }

    @Deprecated(since = "1.20")
    @Override
    public void spawnAfterBreak(@NotNull BlockState state, @NotNull ServerLevel world, @NotNull BlockPos pos, @NotNull ItemStack stack, boolean dropExperience) {
        // Stop Vanilla Stuff
    }

    @Override
    public void animateTick(@NotNull BlockState state, @NotNull Level world, @NotNull BlockPos pos, @NotNull RandomSource random) {
        for (var i = 0; i < (getThickness(state) * 2) + 1; i++)
            world.addAlwaysVisibleParticle(Particles.ACID, pos.getX() + random.nextDouble(),
                    pos.getY() + (Boolean.TRUE.equals(state.getValue(WATERLOGGED)) ? random.nextDouble() : 0.01),
                    pos.getZ() + random.nextDouble(), 0.0, 0.0, 0.0);
        if (random.nextInt(5 * ((MAX_THICKNESS + 1) - getThickness(state))) == 0)
            world.playLocalSound(pos.getX(), pos.getY(), pos.getZ(), SoundEvents.LAVA_EXTINGUISH, SoundSource.BLOCKS,
                    0.2f + random.nextFloat() * 0.2f, 0.9f + random.nextFloat() * 0.15f, false);
    }

    @Deprecated(since = "1.20")
    @Override
    public void tick(@NotNull BlockState state, @NotNull ServerLevel level, @NotNull BlockPos pos, @NotNull RandomSource random) {
        var currentThickness = getThickness(state);
        var canGrief = level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
        if (level.getBlockState(pos.above()).is(GigBlocks.ACID_BLOCK) && currentThickness < 4)
            this.setThickness(level, pos, state, currentThickness + 1);
        if (random.nextInt(8 - currentThickness) == 0 && currentThickness >= 1) {
            if (currentThickness < 4) setThickness(level, pos, state, Math.min(4, level.getRandom().nextInt(3) + 1));
            if (canGrief && !level.getBlockState(pos.below()).is(GigTags.ACID_RESISTANT)) {
                level.setBlock(pos.below(), Blocks.AIR.defaultBlockState(), 2);
                level.playLocalSound(pos.getX(), pos.getY(), pos.getZ(), SoundEvents.LAVA_EXTINGUISH,
                        SoundSource.BLOCKS, 0.2f + random.nextFloat() * 0.2f, 0.9f + random.nextFloat() * 0.15f, false);
            }
        }

        super.tick(state, level, pos, random);
        scheduleTickIfNotScheduled(level, pos);
    }

    @Deprecated(since = "1.20")
    @Override
    public void randomTick(@NotNull BlockState state, @NotNull ServerLevel world, @NotNull BlockPos pos, @NotNull RandomSource random) {
        scheduleTickIfNotScheduled(world, pos);
    }

    @Deprecated(since = "1.20")
    @Override
    public @NotNull VoxelShape getShape(@NotNull BlockState state, @NotNull BlockGetter view, @NotNull BlockPos pos, @NotNull CollisionContext context) {
        return Block.box(0, 0, 0, 16, 2, 16);
    }

    @Deprecated(since = "1.20")
    @Override
    public void attack(@NotNull BlockState state, @NotNull Level world, @NotNull BlockPos pos, @NotNull Player player) {
        // Stop Vanilla stuff
    }

    @Deprecated(since = "1.20")
    @Override
    public boolean canSurvive(@NotNull BlockState state, LevelReader world, @NotNull BlockPos pos) {
        return world.getBlockState(pos).isAir() && !world.getBlockState(pos).is(Blocks.LAVA);
    }

    @Deprecated(since = "1.20")
    @Override
    public void entityInside(@NotNull BlockState blockState, @NotNull Level level, @NotNull BlockPos blockPos, @NotNull Entity entity) {
        if (entity instanceof LivingEntity livingEntity) {
            if (livingEntity.hasEffect(GigStatusEffects.ACID)) return;
            if (livingEntity.getType().is(GigTags.ACID_RESISTANT_ENTITY)) return;
            var itemStack = livingEntity.getItemBySlot(EquipmentSlot.FEET);
            if (!(livingEntity instanceof Player) || (livingEntity instanceof Player playerEntity && !(playerEntity.isCreative() || playerEntity.isSpectator()))) {
                livingEntity.addEffect(new MobEffectInstance(GigStatusEffects.ACID, 60, this.getThickness(blockState)));
            }
            if (!itemStack.equals(ItemStack.EMPTY) && !itemStack.is(GigTags.ACID_RESISTANT_ITEMS)) {
                itemStack.setDamageValue(
                        itemStack.getDamageValue() + (level.getRandom().nextInt(2) * this.getThickness(blockState)));
            }
        }
        if (entity instanceof ItemEntity itemEntity && level.getRandom().nextInt(20) < 2) {
            if (itemEntity.getItem().is(GigTags.ACID_RESISTANT_ITEMS)) return;
            if (itemEntity.getItem().getMaxDamage() < 2) {
                itemEntity.getItem().shrink(1);
            } else {
                itemEntity.getItem().setDamageValue(
                        itemEntity.getItem().getDamageValue() + (level.getRandom().nextInt(2) * this.getThickness(
                                blockState)));
            }
        }
    }
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/block/AcidBlock.kt
package com.bvanseg.gigeresque.common.block

import com.bvanseg.gigeresque.client.particle.Particles
import com.bvanseg.gigeresque.common.entity.AlienEntity
import com.bvanseg.gigeresque.common.extensions.isAcidResistant
import com.bvanseg.gigeresque.common.source.DamageSources
import com.bvanseg.gigeresque.common.util.clamp
import net.minecraft.block.*
import net.minecraft.block.Blocks
import net.minecraft.entity.Entity
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.entity.boss.WitherEntity
import net.minecraft.fluid.FluidState
import net.minecraft.fluid.Fluids
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.IntProperty
import net.minecraft.state.property.Properties
import net.minecraft.tag.BlockTags
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.GameRules
import net.minecraft.world.World
import net.minecraft.world.WorldAccess
import java.util.*
import kotlin.math.max

/**
 * @author Boston Vanseghi
 */
class AcidBlock(settings: Settings) : CustomFallingBlock(settings), Waterloggable {

    companion object {
        private const val MAX_THICKNESS = 4

        val THICKNESS: IntProperty = IntProperty.of("thickness", 0, MAX_THICKNESS)
        private val WATERLOGGED: BooleanProperty = Properties.WATERLOGGED
    }

    init {
        defaultState = (stateManager.defaultState.with(WATERLOGGED, false)).with(THICKNESS, MAX_THICKNESS)
    }

    private fun scheduleTickIfNotScheduled(world: World, pos: BlockPos) {
        if (!world.blockTickScheduler.isScheduled(pos, this)) {
            world.blockTickScheduler.schedule(pos, this, 40)
        }
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(THICKNESS, WATERLOGGED)
    }

    override fun onBlockAdded(state: BlockState, world: World, pos: BlockPos, oldState: BlockState, notify: Boolean) {
        scheduleTickIfNotScheduled(world, pos)
    }

    override fun hasRandomTicks(state: BlockState): Boolean = true
    override fun canMobSpawnInside(): Boolean = false
    override fun getBlastResistance(): Float = 100f
    override fun getRenderType(state: BlockState): BlockRenderType = BlockRenderType.INVISIBLE
    override fun canPathfindThrough(
        state: BlockState,
        world: BlockView,
        pos: BlockPos,
        type: NavigationType
    ): Boolean = true

    override fun getFluidState(state: BlockState): FluidState? {
        return if (state.get(StairsBlock.WATERLOGGED) as Boolean) Fluids.WATER.getStill(false) else super.getFluidState(
            state
        )
    }

    override fun getStateForNeighborUpdate(
        state: BlockState,
        direction: Direction,
        neighborState: BlockState,
        world: WorldAccess,
        pos: BlockPos?,
        neighborPos: BlockPos?
    ): BlockState {
        if (state.get(WATERLOGGED) as Boolean) {
            world.fluidTickScheduler.schedule(pos, Fluids.WATER, Fluids.WATER.getTickRate(world))
        }
        return super.getStateForNeighborUpdate(state, direction, neighborState, world, pos, neighborPos)
    }

    private fun getThickness(state: BlockState): Int = state.get(THICKNESS) as Int
    private fun setThickness(world: ServerWorld, pos: BlockPos, state: BlockState, consume: Int = 1) {
        val newThickness = max(getThickness(state) - consume, 0)

        var newState = state.with(THICKNESS, newThickness)

        if (world.getBlockState(pos).block == Blocks.WATER) {
            newState = newState.with(WATERLOGGED, true)
        }

        world.setBlockState(pos, newState)
    }

    override fun scheduledTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        val currentThickness = getThickness(state)

        if (currentThickness <= 0) {
            world.breakBlock(pos, false)
            return
        }

        if (random.nextInt(8 - currentThickness) == 0) {
            val canGrief = world.gameRules.getBoolean(GameRules.DO_MOB_GRIEFING)

            val blockToEat = pos.down()

            if (currentThickness >= 1) {
                setThickness(world, pos, state, clamp(random.nextInt(2) + 1, 1, currentThickness))

                if (canGrief && !world.getBlockState(blockToEat).block.isAcidResistant) {
                    world.breakBlock(blockToEat, false)
                } else {
                    world.breakBlock(pos, false)
                }

            }
        }

        super.scheduledTick(state, world, pos, random)
        scheduleTickIfNotScheduled(world, pos)
    }

    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random?) {
        scheduleTickIfNotScheduled(world, pos)
    }

    override fun randomDisplayTick(state: BlockState, world: World, pos: BlockPos, random: Random) {
        if (getThickness(state) == 0) {
            world.breakBlock(pos, false)
            return
        }

        for (i in 0 until (getThickness(state) * 2) + 1) {
            val yOffset = if (state.get(WATERLOGGED) as Boolean) random.nextDouble() else 0.01
            val d = pos.x.toDouble() + random.nextDouble()
            val e = pos.y.toDouble() + yOffset
            val f = pos.z.toDouble() + random.nextDouble()
            world.addImportantParticle(Particles.ACID, d, e, f, 0.0, 0.0, 0.0)
        }
        if (random.nextInt(5 * ((MAX_THICKNESS + 1) - getThickness(state))) == 0) {
            world.playSound(
                pos.x.toDouble(),
                pos.y.toDouble(),
                pos.z.toDouble(),
                SoundEvents.BLOCK_LAVA_EXTINGUISH,
                SoundCategory.BLOCKS,
                0.2f + random.nextFloat() * 0.2f,
                0.9f + random.nextFloat() * 0.15f,
                false
            )

        }
    }

    override fun canFallThrough(state: BlockState): Boolean {
        val material = state.material
        return (state.isAir || state.isIn(BlockTags.FIRE) || material.isReplaceable) && !material.isLiquid
    }

    override fun getOutlineShape(
        state: BlockState,
        world: BlockView,
        pos: BlockPos,
        context: ShapeContext
    ): VoxelShape = VoxelShapes.empty()

    private fun dealAcidDamage(state: BlockState, entity: Entity) {
        if (entity !is AlienEntity && entity !is WitherEntity) {
            entity.damage(DamageSources.ACID, getThickness(state).toFloat())
        }
    }

    override fun onEntityCollision(state: BlockState, world: World, pos: BlockPos, entity: Entity) {
        dealAcidDamage(state, entity)
    }

    override fun onSteppedOn(world: World, pos: BlockPos, state: BlockState, entity: Entity) {
        dealAcidDamage(state, entity)
    }
}